### ISSUE TITLE
Overload implementation not checked against overload signatures

### ISSUE THREAD

**Issue by KevOrr** [microsoft/pyright#2806]
**Describe the bug**
When typechecking an overload implementation, the body is checked only in the context of the implementation's signature, and not each overload's signature. This leads to false positives and false negatives.

**To Reproduce**
Run pyright on this snippet:
```py
from typing import Union, overload

@overload
def f(a: str, b: str) -> str:
    ...

@overload
def f(a: bytes, b: bytes) -> bytes:
    ...

def f(a: Union[str, bytes], b: Union[str, bytes]) -> Union[str, bytes]:
    return a + b
```

Replacing the overload implementation with the following also results in a similar error:

```py
def f(a: Union[str, bytes], b: Union[str, bytes]) -> Union[str, bytes]:
    if isinstance(a, str):
        return a + b
    elif isinstance(a, bytes):
        return a + b
```

Replacing the implementation with the following results in no errors (as expected):

```py
def f(a: Union[str, bytes], b: Union[str, bytes]) -> Union[str, bytes]:
    if isinstance(a, str) and isinstance(b, str):
        return a + b
    elif isinstance(a, bytes) and isinstance(b, bytes):
        return a + b
    else:
        assert False
```

Replacing the implementation with the following results in no errors, although I think this should be an error (note how the return type in each case is swapped):

```py
def f(a: Union[str, bytes], b: Union[str, bytes]) -> Union[str, bytes]:
    if isinstance(a, str) and isinstance(b, str):
        return b'bytes'
    elif isinstance(a, bytes) and isinstance(b, bytes):
        return 'str'
    else:
        assert False
```

**Expected behavior**
I would expect pyright to typecheck the implementation body in the context of each overload signature. So first `a: str, b: str |- a + b : str` would be checked, and then `a: bytes, b: bytes |- a + b : bytes` would be checked. Instead, it seems like the body is typechecked in the context of the implementation's overall signature, which forgets the relationship between the parameters' types and the return type.

**VS Code extension or command-line**
CLI and as an LSP server. 1.1.203

**Comment by erictraut** [comment]
This is the correct implementation. The body of the implementation is checked only against the implementation signature. The overload signatures are confirmed to be "proper subsets" of the implementation signature. The implementation is not checked against the individual overloads. In general, that's not possible because the overload signatures can omit parameters and otherwise restrict types in a way that wouldn't make sense for the implementation.

Pyright's behavior here is consistent with other type checkers like mypy.

**Comment by erictraut** [comment]
In this particular example, you should probably be using a constrained TypeVar rather than overloads.

```python
def f(a: AnyStr, b: AnyStr) -> AnyStr:
    return a + b
```

**Comment by KevOrr** [comment]
Yes, in this example, a type var is more suitable. This is just a MRE of a more complex example.

Interesting, are there any examples where checking against an overload would not make sense?

**Comment by erictraut** [comment]
It's typically not possible to check against an overload signature. Look at most of the overloads in the typeshed stubs, for example.

Here's a simple, contrived example where the first overload omits parameters.
```python
@overload
def f() -> str: ...

@overload
def f(a: int) -> int: ...

@overload
def f(a: str) -> str: ...

def f(a: int | str | None = None) -> int | str:
    if a is None:
        return "None"
    return a
```

**Comment by KevOrr** [comment]
Ah, I see. That's more complex than I was originally thinking. Still, when typechecking against the first overload, I would think the type of the default argument would kick in, so the context would be `a : None`. Although I'm in over my head once type variance gets involved, so not sure what would be done if you have a `def f()` overload and a say `def f(a: list[int] | list[str] = [])` implementation. Unfortunate, but understandable if there are tough questions like that.

As an aside, in preferring parametricity over overloads in cases like these, is it possible to make this well-typed:
```py
def f(s: AnyStr, x: AnyStr = None):
    if x is None:
        if isinstance(s, str):
            x = '...'
        else:
            x = b'...'
    ...
```
?

**Comment by erictraut** [comment]
That will type check fine in mypy because it does multiple analysis passes for functions that use constrained TypeVars, but it won't type check in pyright because multiple analysis passes are not feasible in a language server. (There are many other downsides to multiple analysis passes like performance overhead and multiple redundant/confusing error messages.)

I recently added support for return expressions that depend on narrowed constrained TypeVars, so the following now works.

```python
def f(s: AnyStr) -> AnyStr:
    if isinstance(s, str):
        return '...'
    else:
        return b'...'
```

This was a very complex and difficult feature to add. I don't think I can extend it to arbitrary assignments of "linked" variables (i.e. variables that use the same constrained TypeVar in their type declaration).

**Comment by KevOrr** [comment]
Oh, that might be enough to abstract out such uses. Thanks for your help
